File: solutions/213.house-robber-ii.java
/*
 * @lc app=leetcode id=213 lang=java
 *
 * [213] House Robber II
 */

// @lc code=start
class Solution {
    public int rob(int[] nums) {
        if (nums.length == 0) return 0;
        if (nums.length == 1) return nums[0];

        return Math.max(rob(nums, 0, nums.length - 2), rob(nums, 1, nums.length - 1));
    }

    private int rob(int[] nums, int begin, int end) {
        int prevHouse = 0;
        int lastHouse = 0;
        for(int i = begin; i <= end; i++) {
            int temp = Math.max(lastHouse, prevHouse + nums[i]);
            prevHouse = lastHouse;
            lastHouse = temp;
        }
        return lastHouse;
    }
}
// @lc code=end

File: solutions/213.house-robber-ii.py
#
# @lc app=leetcode id=213 lang=python3
#
# [213] House Robber II
#

# @lc code=start
from typing import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 0:
            return 0
        if len(nums) == 1:
            return nums[0]

        def rob(begin, end):
            prev_house, last_house = 0, 0
            for i in range(begin, end + 1):
                prev_house, last_house = last_house, max(last_house, prev_house + nums[i])
            return last_house

        return max(rob(0, len(nums) - 2), rob(1, len(nums) - 1))

# @lc code=end

